Trustroots/trustroots
# Add react tests

~Initial playing around with adding tests for react components. Just a quick play for now...~

This adds react testing in line with the philosophy outlined in the [comment below](https://github.com/Trustroots/trustroots/pull/1147#issuecomment-565741145).

Currently, it means we have two test runners, the existing karma one, and a new jest one. Over time we need to put karma in fire, and just use jest for the client tests.

So, there are two test path patterns now:
* karma tests --> `modules/*/tests/client/*.js` (configured in `karma.conf.js` in `preprocessors` and `files` configuration)
* jest tests --> `modules/*/tests/client/components/*.test.js` (configured in `package.json` at path `jest.testMatch`)

Slightly related, I also fixed the `$templateCache` 404 errors in the tests. I think it had a broken setup for quite a while, the `ngHtml2JsPreprocessor` wasn't even being called (as no html files included in the _files_ section), but we also don't need/use that now, so I removed it. Templates are built in the gulp configuration (called before running the tests) and loaded from `dist/templates.js` (via `main.js`). So they are just always there. With the real templates, not empty strings.

## New/changed commands

```
# run both suites, first karma, then jest
npm run test:client

# run and watch (only) the karma tests
npm run test:client:watch

# run the jest tests
npx jest 

# run and watch jest tests
npx jest --watchAll

# run one specific jest test file
npx jest modules/core/tests/client/components/LanguageSwitch.test.js 

# run and watch one specific jest test file
npx jest --watch modules/core/tests/client/components/LanguageSwitch.test.js 
```

(in the future might put the jest command behind an npm command too...)

## How to review/test

* `git checkout react-tests` (or `git switch react-tests` if you're into the new command, since git v2.23, see https://git-scm.com/docs/git-switch)
* run full client test suite: `npm run test:client` (or full backend too if you want `npm run test`)
* run jest tests: `npx jest`
* optionally try with watching jest tests: `npx jest --watchAll`
* have a peek at how the tests look...
* make _at least_ one important|nitpicky comment for me to change something

## TODO

- [x] investigate https://github.com/testing-library/react-testing-library
- [x] include react tests in main test suite

## TODO later (in other PRs)

- [ ] migrate existing react enzyme tests to this new approach
- [ ] consider migrating ~mocha~ karma/mocha/jasmine -> jest? see https://github.com/skovhus/jest-codemods
- [ ] test a few more react things
- [ ] add useful docs and links into the docs page
